File: app/api/salones_sociales.py
from datetime import date, timedelta

from app import schemas
from app.database import get_db
from app.security import role_required
from app.services import salones_service
from app.services.cloudinary_service import upload_owner_photo
from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

router = APIRouter(prefix="/salones-sociales", tags=["salones-sociales"])


def _range_defaults(
    fecha_inicio: date | None,
    fecha_fin: date | None,
) -> tuple[date, date]:
    start = fecha_inicio or date.today()
    end = fecha_fin or (start + timedelta(days=45))
    if end < start:
        raise HTTPException(status_code=422, detail="La fecha final no puede ser anterior a la inicial")
    if (end - start).days > 120:
        raise HTTPException(status_code=422, detail="El rango máximo del calendario es de 120 días")
    return start, end


@router.get("/admin/salones", response_model=list[schemas.SalonSocialOut])
def admin_list_salones(
    current_user=Depends(role_required(["admin"])),
    db: Session = Depends(get_db),
):
    return salones_service.list_salones(db, current_user.conjunto_id)


@router.post("/admin/salones", response_model=schemas.SalonSocialOut, status_code=201)
async def admin_create_salon(
    nombre: str = Form(...),
    descripcion: str | None = Form(None),
    capacidad: int = Form(...),
    estado: str = Form("activo"),
    color_calendario: str = Form("#0f766e"),
    precio_sin_aseo_centavos: int = Form(salones_service.DEFAULT_PRECIO_SIN_ASEO_CENTAVOS),
    precio_con_aseo_centavos: int = Form(salones_service.DEFAULT_PRECIO_CON_ASEO_CENTAVOS),
    imagen: UploadFile | None = File(None),
    current_user=Depends(role_required(["admin"])),
    db: Session = Depends(get_db),
):
    if estado not in {"activo", "inactivo"}:
        raise HTTPException(status_code=422, detail="Estado de salón inválido")
    if capacidad <= 0:
        raise HTTPException(status_code=422, detail="La capacidad debe ser mayor a cero")
    if not color_calendario.startswith("#") or len(color_calendario) != 7:
        raise HTTPException(status_code=422, detail="Color de calendario inválido")

    imagen_url = None
    if imagen and imagen.filename:
        if not imagen.content_type or not imagen.content_type.startswith("image/"):
            raise HTTPException(status_code=400, detail="La imagen del salón debe ser una imagen")
        imagen_url = upload_owner_photo(imagen)

    try:
        return salones_service.create_salon(
            db,
            conjunto_id=current_user.conjunto_id,
            nombre=nombre.strip(),
            descripcion=descripcion.strip() if descripcion else None,
            capacidad=capacidad,
            imagen_url=imagen_url,
            estado=estado,
            color_calendario=color_calendario,
            precio_sin_aseo_centavos=precio_sin_aseo_centavos,
            precio_con_aseo_centavos=precio_con_aseo_centavos,
        )
    except IntegrityError as exc:
        raise HTTPException(status_code=409, detail="Ya existe un salón con ese nombre") from exc


@router.put("/admin/salones/{salon_id}", response_model=schemas.SalonSocialOut)
async def admin_update_salon(
    salon_id: int,
    nombre: str | None = Form(None),
    descripcion: str | None = Form(None),
    capacidad: int | None = Form(None),
    estado: str | None = Form(None),
    color_calendario: str | None = Form(None),
    precio_sin_aseo_centavos: int | None = Form(None),
    precio_con_aseo_centavos: int | None = Form(None),
    imagen: UploadFile | None = File(None),
    current_user=Depends(role_required(["admin"])),
    db: Session = Depends(get_db),
):
    salon = salones_service.get_salon(db, current_user.conjunto_id, salon_id)
    if not salon:
        raise HTTPException(status_code=404, detail="Salón no encontrado")

    payload = schemas.SalonSocialUpdate(
        nombre=nombre,
        descripcion=descripcion,
        capacidad=capacidad,
        estado=estado,
        color_calendario=color_calendario,
        precio_sin_aseo_centavos=precio_sin_aseo_centavos,
        precio_con_aseo_centavos=precio_con_aseo_centavos,
    )
    imagen_url = None
    if imagen and imagen.filename:
        if not imagen.content_type or not imagen.content_type.startswith("image/"):
            raise HTTPException(status_code=400, detail="La imagen del salón debe ser una imagen")
        imagen_url = upload_owner_photo(imagen)
    try:
        return salones_service.update_salon(db, salon, payload, imagen_url=imagen_url)
    except IntegrityError as exc:
        raise HTTPException(status_code=409, detail="Ya existe un salón con ese nombre") from exc


@router.get("/admin/calendario", response_model=list[schemas.SalonSocialCalendarDayOut])
def admin_calendario(
    fecha_inicio: date | None = Query(None),
    fecha_fin: date | None = Query(None),
    current_user=Depends(role_required(["admin"])),
    db: Session = Depends(get_db),
):
    start, end = _range_defaults(fecha_inicio, fecha_fin)
    return salones_service.calendario(db, current_user.conjunto_id, start, end)


@router.get("/admin/reservas", response_model=list[schemas.SalonSocialReservaOut])
def admin_reservas(
    current_user=Depends(role_required(["admin"])),
    db: Session = Depends(get_db),
):
    return salones_service.list_reservas(db, current_user.conjunto_id)


@router.post("/admin/bloqueos", response_model=schemas.SalonSocialReservaOut | None)
def admin_bloqueo(
    payload: schemas.SalonSocialBloqueoIn,
    current_user=Depends(role_required(["admin"])),
    db: Session = Depends(get_db),
):
    try:
        bloqueo = salones_service.set_bloqueo_admin(db, current_user.conjunto_id, payload, current_user.username)
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    return salones_service._reserva_out(bloqueo) if bloqueo else None


@router.patch("/admin/reservas/{reserva_id}", response_model=schemas.SalonSocialReservaOut)
def admin_update_reserva(
    reserva_id: int,
    payload: schemas.SalonSocialReservaEstadoIn,
    current_user=Depends(role_required(["admin"])),
    db: Session = Depends(get_db),
):
    reserva = salones_service.get_reserva(db, current_user.conjunto_id, reserva_id)
    if not reserva:
        raise HTTPException(status_code=404, detail="Reserva no encontrada")
    try:
        updated = salones_service.update_estado_reserva_admin(db, reserva, payload, current_user.username)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return salones_service._reserva_out(updated)


@router.get("/propietario/salones", response_model=list[schemas.SalonSocialOut])
def propietario_salones(
    current_user=Depends(role_required(["propietario"])),
    db: Session = Depends(get_db),
):
    return salones_service.list_salones(db, current_user.conjunto_id, only_active=True)


@router.get("/propietario/calendario", response_model=list[schemas.SalonSocialCalendarDayOut])
def propietario_calendario(
    fecha_inicio: date | None = Query(None),
    fecha_fin: date | None = Query(None),
    current_user=Depends(role_required(["propietario"])),
    db: Session = Depends(get_db),
):
    start, end = _range_defaults(fecha_inicio, fecha_fin)
    return salones_service.calendario(
        db,
        current_user.conjunto_id,
        start,
        end,
        only_active_salones=True,
    )


@router.get("/propietario/reservas", response_model=list[schemas.SalonSocialReservaOut])
def propietario_reservas(
    current_user=Depends(role_required(["propietario"])),
    db: Session = Depends(get_db),
):
    if current_user.propietario_id is None:
        raise HTTPException(status_code=404, detail="Cuenta de propietario no vinculada")
    return salones_service.list_reservas(
        db,
        current_user.conjunto_id,
        propietario_id=current_user.propietario_id,
    )


@router.post("/propietario/reservas", response_model=schemas.SalonSocialReservaOut, status_code=201)
def propietario_crear_reserva(
    payload: schemas.SalonSocialReservaCreate,
    current_user=Depends(role_required(["propietario"])),
    db: Session = Depends(get_db),
):
    propietario = current_user.propietario
    if not propietario:
        raise HTTPException(status_code=404, detail="Cuenta de propietario no vinculada")
    try:
        reserva = salones_service.create_reserva_propietario(
            db,
            current_user.conjunto_id,
            propietario,
            payload,
        )
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    return salones_service._reserva_out(reserva)


@router.post("/propietario/reservas/{reserva_id}/pago", response_model=schemas.SalonSocialReservaOut)
async def propietario_reportar_pago(
    reserva_id: int,
    referencia_pago: str | None = Form(None),
    notas: str | None = Form(None),
    comprobante: UploadFile | None = File(None),
    current_user=Depends(role_required(["propietario"])),
    db: Session = Depends(get_db),
):
    reserva = salones_service.get_reserva(db, current_user.conjunto_id, reserva_id)
    if not reserva or reserva.propietario_id != current_user.propietario_id:
        raise HTTPException(status_code=404, detail="Reserva no encontrada")
    comprobante_url = None
    if comprobante and comprobante.filename:
        if not comprobante.content_type or not comprobante.content_type.startswith("image/"):
            raise HTTPException(status_code=400, detail="El comprobante debe ser una imagen")
        comprobante_url = upload_owner_photo(comprobante)
    try:
        updated = salones_service.reportar_pago(db, reserva, comprobante_url, referencia_pago, notas)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return salones_service._reserva_out(updated)


@router.post("/propietario/reservas/{reserva_id}/cancelar", response_model=schemas.SalonSocialReservaOut)
def propietario_cancelar_reserva(
    reserva_id: int,
    motivo: str | None = Form(None),
    current_user=Depends(role_required(["propietario"])),
    db: Session = Depends(get_db),
):
    reserva = salones_service.get_reserva(db, current_user.conjunto_id, reserva_id)
    if not reserva or reserva.propietario_id != current_user.propietario_id:
        raise HTTPException(status_code=404, detail="Reserva no encontrada")
    try:
        updated = salones_service.cancel_reserva_propietario(db, reserva, motivo)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return salones_service._reserva_out(updated)

File: frontend/src/api/salonesSociales.ts
import type {
  SalonSocialCalendarDayOut,
  SalonSocialOut,
  SalonSocialReservaOut,
} from '../types'
import apiClient from './axios'

export function moneySalon(centavos: number): string {
  return `$${Math.round(centavos / 100).toLocaleString('es-CO')}`
}

function appendIfPresent(form: FormData, key: string, value: unknown) {
  if (value !== undefined && value !== null && value !== '') {
    form.append(key, String(value))
  }
}

export interface SalonPayload {
  nombre: string
  descripcion?: string | null
  capacidad: number
  estado: 'activo' | 'inactivo'
  color_calendario: string
  precio_sin_aseo_centavos: number
  precio_con_aseo_centavos: number
  imagen?: File | null
}

export async function adminListarSalones(): Promise<SalonSocialOut[]> {
  const { data } = await apiClient.get<SalonSocialOut[]>('/api/v1/salones-sociales/admin/salones')
  return data
}

export async function adminCrearSalon(payload: SalonPayload): Promise<SalonSocialOut> {
  const form = new FormData()
  appendIfPresent(form, 'nombre', payload.nombre)
  appendIfPresent(form, 'descripcion', payload.descripcion)
  appendIfPresent(form, 'capacidad', payload.capacidad)
  appendIfPresent(form, 'estado', payload.estado)
  appendIfPresent(form, 'color_calendario', payload.color_calendario)
  appendIfPresent(form, 'precio_sin_aseo_centavos', payload.precio_sin_aseo_centavos)
  appendIfPresent(form, 'precio_con_aseo_centavos', payload.precio_con_aseo_centavos)
  if (payload.imagen) form.append('imagen', payload.imagen)
  const { data } = await apiClient.post<SalonSocialOut>('/api/v1/salones-sociales/admin/salones', form)
  return data
}

export async function adminActualizarSalon(id: number, payload: Partial<SalonPayload>): Promise<SalonSocialOut> {
  const form = new FormData()
  appendIfPresent(form, 'nombre', payload.nombre)
  appendIfPresent(form, 'descripcion', payload.descripcion)
  appendIfPresent(form, 'capacidad', payload.capacidad)
  appendIfPresent(form, 'estado', payload.estado)
  appendIfPresent(form, 'color_calendario', payload.color_calendario)
  appendIfPresent(form, 'precio_sin_aseo_centavos', payload.precio_sin_aseo_centavos)
  appendIfPresent(form, 'precio_con_aseo_centavos', payload.precio_con_aseo_centavos)
  if (payload.imagen) form.append('imagen', payload.imagen)
  const { data } = await apiClient.put<SalonSocialOut>(`/api/v1/salones-sociales/admin/salones/${id}`, form)
  return data
}

export async function adminCalendarioSalones(fechaInicio: string, fechaFin: string): Promise<SalonSocialCalendarDayOut[]> {
  const { data } = await apiClient.get<SalonSocialCalendarDayOut[]>('/api/v1/salones-sociales/admin/calendario', {
    params: { fecha_inicio: fechaInicio, fecha_fin: fechaFin },
  })
  return data
}

export async function adminListarReservasSalones(): Promise<SalonSocialReservaOut[]> {
  const { data } = await apiClient.get<SalonSocialReservaOut[]>('/api/v1/salones-sociales/admin/reservas')
  return data
}

export async function adminBloquearFechaSalon(payload: {
  salon_id: number
  fecha: string
  estado: 'disponible' | 'no_disponible' | 'mantenimiento' | 'evento_privado'
  notas?: string
}): Promise<SalonSocialReservaOut | null> {
  const { data } = await apiClient.post<SalonSocialReservaOut | null>('/api/v1/salones-sociales/admin/bloqueos', payload)
  return data
}

export async function adminCambiarEstadoReservaSalon(
  reservaId: number,
  estado: 'confirmado' | 'cancelado' | 'pendiente_pago' | 'pendiente_aprobacion',
  motivo?: string,
): Promise<SalonSocialReservaOut> {
  const { data } = await apiClient.patch<SalonSocialReservaOut>(
    `/api/v1/salones-sociales/admin/reservas/${reservaId}`,
    { estado, motivo },
  )
  return data
}

export async function propietarioListarSalones(): Promise<SalonSocialOut[]> {
  const { data } = await apiClient.get<SalonSocialOut[]>('/api/v1/salones-sociales/propietario/salones')
  return data
}

export async function propietarioCalendarioSalones(fechaInicio: string, fechaFin: string): Promise<SalonSocialCalendarDayOut[]> {
  const { data } = await apiClient.get<SalonSocialCalendarDayOut[]>('/api/v1/salones-sociales/propietario/calendario', {
    params: { fecha_inicio: fechaInicio, fecha_fin: fechaFin },
  })
  return data
}

export async function propietarioReservasSalones(): Promise<SalonSocialReservaOut[]> {
  const { data } = await apiClient.get<SalonSocialReservaOut[]>('/api/v1/salones-sociales/propietario/reservas')
  return data
}

export async function propietarioCrearReservaSalon(payload: {
  salon_id: number
  fecha: string
  incluye_aseo: boolean
  notas?: string
}): Promise<SalonSocialReservaOut> {
  const { data } = await apiClient.post<SalonSocialReservaOut>('/api/v1/salones-sociales/propietario/reservas', payload)
  return data
}

export async function propietarioReportarPagoSalon(payload: {
  reserva_id: number
  comprobante?: File | null
  referencia_pago?: string
  notas?: string
}): Promise<SalonSocialReservaOut> {
  const form = new FormData()
  appendIfPresent(form, 'referencia_pago', payload.referencia_pago)
  appendIfPresent(form, 'notas', payload.notas)
  if (payload.comprobante) form.append('comprobante', payload.comprobante)
  const { data } = await apiClient.post<SalonSocialReservaOut>(
    `/api/v1/salones-sociales/propietario/reservas/${payload.reserva_id}/pago`,
    form,
  )
  return data
}

export async function propietarioCancelarReservaSalon(reservaId: number, motivo?: string): Promise<SalonSocialReservaOut> {
  const form = new FormData()
  appendIfPresent(form, 'motivo', motivo)
  const { data } = await apiClient.post<SalonSocialReservaOut>(
    `/api/v1/salones-sociales/propietario/reservas/${reservaId}/cancelar`,
    form,
  )
  return data
}
